zipline/src/commonMain/kotlin/app/cash/zipline/FlowReference.kt

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.zipline
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.emitAll
+import kotlinx.coroutines.flow.transform
+import kotlinx.serialization.Contextual
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+import okio.use
+
+/** A flow reference is a [Flow] that can be transmitted over a Zipline call. */
+@Serializable
+class FlowReference<T> @PublishedApi internal constructor(
+  @Contextual private val referenceFlowReference: ZiplineReference<ReferenceFlow>
+) {
+  private fun getJsonFlow(): Flow<String> {
+    return channelFlow {
```

Comment:
I believe the use of this will break synchronous emission behavior of a Flow, but maybe we're unable to provide that across the bridge anyway?

A test which would catch this looks like this: https://github.com/cashapp/sqldelight/blob/10b5a6605bf2584dbef414b80ecfa601c1c76701/extensions/coroutines-extensions/src/commonTest/kotlin/com/squareup/sqldelight/runtime/coroutines/QueryAsFlowTest.kt#L37-L43 with a source of `flowOf(1)` or something.

Reply:
When I attempt a non-channel `flow {}` I get this failure:
```
java.lang.Exception: java.lang.IllegalStateException: Flow invariant is violated:
		Emission from another coroutine is detected.
		Child of "coroutine#8":StandaloneCoroutine{Active}@22a13908, expected child of "coroutine#6":BlockingCoroutine{Active}@41f1f8b4.
		FlowCollector is not thread-safe and concurrent emissions are prohibited.
		To mitigate this restriction please use 'channelFlow' builder instead of 'flow'
```
There might be other mitigations but this seemed like the most natural.